src/main/java/TreeNode.java

```diff
@@ -1,23 +1,19 @@
 public class TreeNode {
     public TreeNode left;
     public TreeNode right;
+    public TreeNode parent;
+    public int level;		//Level 0 is root
 
-    /* PRE: Assuming that both TreeNode a and b are children of cur, on first call */
-    public static TreeNode lowestCommonAncestor(TreeNode cur, TreeNode a, TreeNode b) {
-        if(cur == a || cur == b) {
-            return cur;
-        }
-        TreeNode leftTree = lowestCommonAncestor(cur.left, a, b);
-        TreeNode rightTree = lowestCommonAncestor(cur.right, a, b);
-
-        if(leftTree != null && rightTree != null) {    // If both sub-trees have either a or b, current is LCA
-            return cur;
-        }
-
-        if(leftTree != null) {
-            return leftTree;
-        } else {
-            return rightTree;
-        }
+    /* PRE: Assuming that both TreeNode a and b are in the same tree */
+    public static TreeNode lowestCommonAncestor(TreeNode a, TreeNode b) {
+    	while(a != b) {
+    		if(a.level <= b.level) {
+    			a = a.parent;
+    		}
+    		if(a.level >= b.level) {
+    			b = b.parent;
+    		}
+    	}
```

Comment:
This while loop won't work for a number of reasons.

First, on line 10, if a's level is LESS than b's level, you traverse a up its parent? Shouldn't you be traversing b up it's parent, since b is lower down than a? Same is true for line 13.

Next, assuming you do indeed fix that, consider the following tree:

    0
   /  \
  1   a
 /
b

What will your code do?

Reply:
Ugh. Stupid github. Tree looks like this:

{0, {1, {b}, {}}, {a}}

So the root has a left child with value 1, and that node has a left child value b, and no right child. The root's right child is a single node that's a.

Reply:
Totally forgot that levels were higher as we descend, so fixed to reflect!